System/Posix/IO/Common.hsc

```diff
@@ -412,6 +416,20 @@ fdRead fd nbytes = do
        s <- peekCStringLen (castPtr buf, fromIntegral n)
        return (s, n)
 
+-- | Read data from an 'Fd'.
+-- Throws an exception if this is an invalid descriptor, or EOF has been
+-- reached.
+fdReadBytes :: Fd
+       -> ByteCount -- ^How many bytes to read
+       -> IO (ByteString, ByteCount) -- ^The bytes read, how many bytes were read.
```

Comment:
`ByteStrings` already carry a length field, there's little point in returning it another copy wrapped in a tuple.  This should use `createUptoN` and return just the `ByteString`.

Also, not all errors are invalid descriptor or `EOF`. On a non-blocking file descriptor, the error could be EAGAIN, and `fdReadBuf` will have thrown the corresponding IO Exception.
I am not sure that raising on `EOF` is the most useful result, returning an empty ByteString may be a cleaner choice, but perhaps we're too used to EOF exceptions in Haskell... :-(

Reply:
> Also, not all errors are invalid descriptor or `EOF`. On a non-blocking file descriptor, the error could be EAGAIN, and `fdReadBuf` will have thrown the corresponding IO Exception.
> I am not sure that raising on `EOF` is the most useful result, returning an empty ByteString may be a cleaner choice, but perhaps we're too used to EOF exceptions in Haskell... :-(

Not sure yet what my opinion is here; I tried to stick to the existing design (with all its bugs and quirks) here. If we decide to change the way how we handle EOF I suggest we do it in https://github.com/haskell/unix/pull/219 and discuss it over there.